java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,520 @@
+          internalTransferPairs[i].transfer();
+        }
+      }
+      to.valueCount = valueCount;
+      clear();
+    }
+
+    @Override
+    public void splitAndTransfer(int startIndex, int length) {
+      to.clear();
+      int startPoint = startIndex * TYPE_WIDTH;
+      int sliceLength = length * TYPE_WIDTH;
+      ArrowBuf slicedBuffer = typeBuffer.slice(startPoint, sliceLength);
+      ReferenceManager refManager = slicedBuffer.getReferenceManager();
+      to.typeBuffer = refManager.transferOwnership(slicedBuffer, to.allocator).getTransferredBuffer();
+
+      while (to.offsetBuffer.capacity() < length * OFFSET_WIDTH) {
+        to.reallocOffsetBuffer();
+      }
+
+      int [] typeCounts = new int[MinorType.values().length];
+      int [] typeStarts = new int[Types.MinorType.values().length];
+      for (int i = 0; i < typeCounts.length; i++) {
+        typeCounts[i] = 0;
+        typeStarts[i] = -1;
+      }
+
+      for (int i = startIndex; i < startIndex + length; i++) {
+        int ord = typeBuffer.getByte(i);
+        to.offsetBuffer.setInt((i - startIndex) * OFFSET_WIDTH, typeCounts[ord]);
+        typeCounts[ord] += 1;
+        if (typeStarts[ord] == -1) {
+          typeStarts[ord] = offsetBuffer.getInt(i * OFFSET_WIDTH);
+        }
+      }
+      to.setValueCount(length);
+
+      // transfer vector values
+      for (int i = 0; i < MinorType.values().length; i++) {
+        if (typeCounts[i] > 0 && typeStarts[i] != -1) {
+          internalTransferPairs[i].splitAndTransfer(typeStarts[i], typeCounts[i]);
+        }
+      }
+    }
+
+    @Override
+    public ValueVector getTo() {
+      return to;
+    }
+
+    @Override
+    public void copyValueSafe(int from, int to) {
+      this.to.copyFrom(from, to, DenseUnionVector.this);
+    }
+  }
+
+  @Override
+  public FieldReader getReader() {
+    if (reader == null) {
+      reader = new UnionReader(this);
+    }
+    return reader;
+  }
+
+  public FieldWriter getWriter() {
+    if (writer == null) {
+      writer = new UnionWriter(this);
+    }
+    return writer;
+  }
+
+  @Override
+  public int getBufferSize() {
+    if (valueCount == 0) { return 0; }
+
+    return valueCount * TYPE_WIDTH + valueCount * OFFSET_WIDTH + internalStruct.getBufferSize();
+  }
+
+  @Override
+  public int getBufferSizeFor(final int valueCount) {
+    if (valueCount == 0) {
+      return 0;
+    }
+
+    long bufferSize = 0;
+    for (final ValueVector v : (Iterable<ValueVector>) this) {
+      bufferSize += v.getBufferSizeFor(valueCount);
+    }
+
+    return (int) bufferSize + valueCount * TYPE_WIDTH + valueCount * OFFSET_WIDTH;
+  }
+
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
+    List<ArrowBuf> list = new java.util.ArrayList<>();
+    setReaderAndWriterIndex();
+    if (getBufferSize() != 0) {
+      list.add(typeBuffer);
+      list.add(offsetBuffer);
+      list.addAll(java.util.Arrays.asList(internalStruct.getBuffers(clear)));
+    }
+    if (clear) {
```

Comment:
can this call clear instead?

Reply:
yes, I think so. Calling clear/close is essentially the same as calling release, which decrements the refcount by 1.